### PR TITLE
Standardize all commands with autocomplete, help output, and tests

### PR DESCRIPTION
**Pull request by sethvargo**
Okay, here we go! This PR refactors and standardizes all commands to have autocompletion, subcommand completion, predictions based on existing paths, and significantly more test coverage and edge case testing.

High level features:

- New colored UI shows errors in red, warnings in yellow/orange, etc.

- Some commands can auto-complete existing Vault data. For example, `vault read se<tab>` will likely complete to `vault read secret/`, if the user has the proper permissions. This autocompletion applies to mounts, policies, auths, and many other things. In all situations, if communication fails or times out, the command returns empty, so the user will never see a failure (they just might not get their auto-completion).

- New flag parsing library allows for customizing help output in a more consistent way, sharing flags, and automatically printing help output. That's right - all flag-related help output is auto-generated. Output can even be grouped in the display.

- Almost every command now features a full set of unit and integration tests. Also, the tests can be fully parallelized and run in ~15s on my mac.

- Many inconsistencies have been removed, but everything is backwards-compatible.

- Deprecations are clearly marked with a TODO

- Added many convenience functions for consistent output, etc.

- Detect if we are piping to another process and automatically use a blank writer and strip newlines.

- New `auth-list` and `auth-help` commands to replace the confusing `auth -methods` and `auth -method=foo -method-help` (which wasn't at all confusing...). This is fully backwards compatible.

There are a few things up for discussion, and I'd like feedback.

### CODE REVIEW

File: api/secret.go [hunk 1]
```diff
@@ -35,6 +37,268 @@ type Secret struct {
...
+	}
+
+	return accessor
+}
+
+// TokenMeta returns the standardized token metadata for the given secret.
+// If the secret is nil or does not contain an accessor, this returns the empty
+// string. Metadata is usually modeled as an map[string]interface{}, but token
+// metadata is always a map[string]string. This function handles the coercion.
+func (s *Secret) TokenMeta() map[string]string {
```
Comment (jefferai) [inline api/secret.go]: This seems redundant with TokenMetadata below, Can we just use TokenMetadata and have it autodetect? Currently nothing will ever return both an Auth and Data so it should work in all situations.

File: api/secret.go [hunk 2]
```diff
@@ -35,6 +37,268 @@ type Secret struct {
...
+	renewable, ok := s.Data["renewable"].(bool)
+	if !ok {
+		return false
+	}
+
+	return renewable
+}
+
+// TokenTTLInt returns the token's TTL as an integer number of seconds.
+func (s *Secret) TokenTTLInt() int {
```
Comment (jefferai) [inline api/secret.go]: I would get rid of this in favor of the call below. time.Duration is a standard way to do durations in Go code and anyone that needs the number of seconds can get it directly from that.

File: builtin/credential/okta/cli.go [hunk 3]
```diff
@@ -53,14 +53,28 @@ func (h *CLIHandler) Auth(c *api.Client, m map[string]string) (*api.Secret, erro
...
 func (h *CLIHandler) Help() string {
 	help := `
-The Okta credential provider allows you to authenticate with Okta.
-To use it, first configure it through the "config" endpoint, and then
-login by specifying username and password. If password is not provided
-on the command line, it will be read from stdin.
+Usage: vault login -method=okta [CONFIG K=V...]
 
-    Example: vault auth -method=okta username=john
+  The OKTA auth method allows users to authenticate using OKTA.
```
Comment (jefferai) [inline builtin/credential/okta/cli.go]: s/OKTA/Okta

File: api/secret.go (hunk 2 above)
Comment (sethvargo) [reply api/secret.go]: This is here because LeaseDuration is always specified as seconds and often stored as an `int` in some places. time.Duration.Seconds() returns a float, which then has to be rounded and converted to seconds on each call.

**Comment by jefferai** [comment]
And away

**Comment by jefferai** [comment]
We

**Comment by sethvargo** [comment]
🎉
